### PR TITLE
Add NOAA availability function

### PR DESCRIPTION
**Pull request by echaidemenos**
This PR resolves https://github.com/aqualinkorg/aqualink-app/issues/784

Adds a python script for extracting data availability maps from netcdf4 files acquired from https://coralreefwatch.noaa.gov/

`getNOAANearestAvailablePoint` maps (lon, lat) to nearest available point of the noaa api, running a breadth first search on the 7200 by 3200 world availability map matrix

### CODE REVIEW

**Comment by ericboucher** [comment]
@DanielJDufour I am surprised that the tests are failing and the values would change significantly after clipping to the nearest point. Could you have a quick look and make sure this latest change makes sense? Thanks

**Comment by DanielJDufour** [comment]
Hello.  I hope everyone had a good weekend and good start to the week. I looked over the changes and researched the data.  I believe this is the dataset that is being used https://coralreefwatch.noaa.gov/product/5km/index_5km_dhw.php.  As you can see that the raster values appear to change significantly within relatively short distances.  In general, it's been my experience that temperature data is often low resolution and fairly continuous with gradual changes between areas.  However, I think this current dataset we are using breaks from this mold because it's using percentages and not absolute values.  Additionally, the data appears to be at 5km resolution.  This is relatively good resolution for temperature data.

It's a hypothesis/guess, but what I think this means for our data is that the availability grid is at too low of a resolution (i.e. too much distance between points/cells) for simple snapping with type of higher-resolution dataset.  Here's a couple paths forward:
- Improve snapping algorithm in 2 ways: (1) adding check if snapping is necessary, (2) snapping to grid cell edges.   (1) It's currently just snapping to the middle of the current or closest available grid cell.  It's doing this even if it's already with a valid area.  We could add an if statement to check if the current lat/long is within the square cell area around the closest valid pixel.  Then only proceed with saving the snapping if it's not within a valid area.  (2) Because it's simply snapping to the center of the availability grid cells, it's snapping past the edges of the grid cell to the center.  That's relatively a lot of distance for our dataset.  In this case, we could improve the snapping algorithm, to snap to the edge of the grid cell.  There might be floating point arithmetic issues, so it might be advisable to snap slightly inside of the available grid cell.
- Resample the availability grid to a higher resolution (it'll lead to a higher bundle size).
- Bake a GeoTIFF of the availability grid into the code.  Would add significant complexity.
- Compress availability grid another way.  Would also add complexity

This is only a hypothesis, so I think the next steps would be to download the raw data from NOAA and compare test points with their snapped point.

Screen Shot 2022-10-10 at 10 56 30 PM

File: packages/api/src/utils/sofar-availability.ts [hunk 1]
```diff
@@ -18,5 +19,38 @@ export function getSofarNearestAvailablePoint(point: Coord): [number, number] {
...
+        [
+          ((180 + longitude - 0.25) % 360) - 180,
+          ((90 + latitude + 0.25) % 180) - 90,
+        ],
+      ],
+    ],
+  };
+
+  return booleanPointInPolygon(point, poly)
+    ? ((point as Point).coordinates as [number, number])
```
Comment (ericboucher) [inline packages/api/src/utils/sofar-availability.ts:54]: This line won't work if `point` is of shape [number, number] already and not actually a `Point`.

File: packages/api/scripts/netcdfToAvailabilityBitArray.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ds = Dataset(fn)
+# print(ds)
+
+# inset some valid variable. variables and additional information can bee seen with the above command: print(ds)
+var = ds['degree_heating_week'][:]
+
+world = []
+
+for i in range(7200):
+  world.append(var[0, 0:3600, i].mask)
```
[the remark this answers is not in the document]
Comment (ericboucher) [reply packages/api/scripts/netcdfToAvailabilityBitArray.py]: @echaidemenos

**Comment by ericboucher** [comment]
Replaced by #791

**Pull request closed without merging**